### ISSUE TITLE
New Liquidity Mining Strategy

### ISSUE THREAD

**Issue by Nullably** [CoinAlpha/hummingbot#2669]
**Feature Suggestion**
This new strategy allows user to run market making on markets that are yielding them liquidity mining rewards. Key unique features that this strategy offers:
* Diversification by running a market making strategy across many campaigns.
  Diversification helps achieve smoother return
* Better capital allocation by sizing orders based on target asset allocation.
  This facilitates better and full usage of user assets. 
* Dynamic spread adjustment based on market volatility.
  Fixed default spread (currently 1% on PMM) can be easily gamed. Highly volatile markets need wider spreads.
* Easy setup and maintenance.
  A better new users on boarding experience, the strategy will also allow for easier strategy management (in comparison to running/managing multiple bots) and offer automation in markets participation and assets allocation.

*Configuration Parameters*
* exchange
* auto_campaigns_participation
  A flag to indicate whether to automatically participate in as many mining campaign as possible based on token amount the user has in his/her account. The strategy keeps track of and participates in only active LM markets. It stops market making on ended campaigns and starts market making on newly added campaigns automatically. 
* market_making_assets
  To pick what assets to market make, either base or quote, the bot will automatically select eligible markets.
* markets
  A list of fixed LM markets, only in use when `auto_markets_allocation` is `false`. When configured, the bot retrieves a list of current LM campaigns and their rewards from https://miner.hummingbot.io/ (this can be done through API) and for each market prompt asks user whether they want to participate in.
* reserved_balances
  A list of token and the reserved balance (to not be used by the strategy). For example, if a user has 1 BTC in balance and sets balancer reserve to 0.75, this means the balance available to the strategy is 0.25 BTC (0.75 is reserved). This can be used for an asset amount that you want to set a side to pay for fee (e.g. BNB for Binance) or for an amount that you don't want the strategy to use at all (to limit exposure or in anticipation of withdrawal).
* auto_assign_campaign_budgets
  A flag to indicate whether to automatically assign budgets (equal weight) for all campaign markets. The assignment assumes full allocation of user assets (after accounting for reserved balance). Campaign buy budgets are allocated from quote asset balance, e.g. if user has 100 USDT and participates in 4 USDT campaign, each campaign will get buy budget of 25 USDT. Campaign sell budgets are assigned from base asset balance, e.g. if user has 1000 XEM and participates in 2 XEM campaign, each campaign will get sell budget of 500 XEM.
* campaign_budgets
  A list of campaign (market) and their buy and sell budgets, only in used when `auto_assign_campaign_budgets` is not set. 
  For example, `XEM-ETH:500-2, XEM-USDT:300-250` means on XEM-ETH campaign the bot assigns 500 XEM budget for a sell order and 2 ETH budget for a buy order, on XEM-USDT the bot assigns 300 XEM budget for a sell order and 250 USDT budget for a buy order. Budgets can shift between buy and sell orders, for example once a XEM buy order of 2 ETH is filled, the sell budget increased by the XEM amount bought, while buy budget is depleted by 2 ETH.  
* spread_level (tight/medium/wide/custom)
  The level of spread as a percentage to max spread eligible for liquidity mining. 
  `tight` = 25%, `medium` = 50%, `wide` = 75%
  For example, for ALGO-BTC the max spread for LM is 2%, `tight` then means 0.5% spread, `medium` means 1% and `wide` means 1.5%
  The lower the spread, the higher the reward but the higher the chance of achieving poor trade PnL in highly volatile market condition.
* custom_spread_pct
  The spread (for both bid and ask) in use when spread_level is set to `custom`

  
*Other parameters (not prompted on create)*

* order_refresh_time (default to 10)
* order_refresh_tolerance (default to 0.2)
* LM_refresh_time (default to 1800)
  The time interval (in second) to poll a list of LM markets and their yields to auto adjust market participation, in use when `auto_markets_participation` is set.
* spread_volatility_multiple
  Spreads are adjusted by applying multiple to volatility, higher volatility will result in wider spread (where initial spread is the minimum). The volatility can be computed from average price movement over a certain period of time, see /scripts/spreads_adjusted_on_volatility_script.py for more details.

*Nice-to-have parameters*

* fixed_budget_usd
  To specify budget size in USD amount applied to all markets, the budget is then split into buy and sell budgets according to user's balances, first allocated sell budget (to an equivalent amount in USD) then if there is left over assign that to buy budget.
* allocation_based_on_yield
  A flag to indicate whether to automatically assign campaign budgets based on yield, the higher the yield the higher budgets are allocated to the campaigns. Only in used when `auto_assign_campaign_budgets` is set.
* auto_sell_on_ended_campaigns
  A flag to indicate whether to sell the underlying tokens on ended campaigns, this is to roll over capital to participate in existing or new campaigns. Only in use if `auto_markets_participation` is set.
* auto_adjust_quote_asset_balance
  A flag to indicate whether to automatically adjust quote asset balance before starting the strategy. For example, user might have just USDT in their account but wants to also participate in BTC markets, the bot will buy BTC (with USDT) to achieve desired amount to start market making. 

*Design*
* This is cut down version of PMM features with extra functionality on multiple markets making and capital allocation.
* It is designed to be easy to use for most new and advanced users, simply answer yes to most `auto` parameter questions and the bot does the rest for the user. 
* The `create liquidity_mining` command will guide user through creating a new strategy nicely by integrating both user balance information and current active mining campaigns into the steps. This creates more personalised approach. The user needs to `connect` to the exchange before `create` this strategy. 
* The strategy runs a connector which loads up and maintains multiple order books. 

*Limitation*
* It first supports only single order level for ease of management and staying within API rate limit.
* Not sure if we need to open up script support for this.

**Comment by mifeng** [comment]
@Nullably do you think this will make it into v0.36?

**Comment by alexandrucoman** [comment]
I am waiting for it :)

**Comment by Nullably** [comment]
Please refer to this document for actual scope of the first version of this new strategy.
https://www.notion.so/hummingbot/Liquidity-Mining-strategy-0c6eb43ac5ba4c4685e3092f404e0492
